### PR TITLE
Add code action to implement missing trait members

### PR DESCRIPTION
**Pull request by integraledelebesgue**
Closes  [starkware-libs/cairo#5944](https://github.com/starkware-libs/cairo/issues/5944#issue-2386254154)

## Changes
* If `impl` block is missing some members of an implemented trait, a quick fix filling those members appears
* Known generic arguments are inserted in impls, unknown substituted with placeholders
* Inserted code is not formatted

### CODE REVIEW

File: tests/test_data/code_actions/fill_trait_members.txt [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+fn generic<const V: core::integer::u32, W, +Into<core::integer::u32, W>>(w: W) {}
+
+fn with_concrete_impl<W, impl SomeImpl: Into<core::integer::u32, W>>(w: W) -> W {}"
+At: Range { start: Position { line: 31, character: 47 }, end: Position { line: 31, character: 47 } }
+
+//! > Code action #3
+    impl ImplWithEverything<caret> of MyTrait<u32, felt252, 12> {
+Title: Implement missing members
+Add new text: ""
```
Comment (integraledelebesgue) [inline tests/test_data/code_actions/fill_trait_members.txt:137]: @piotmag769 An empty string shouldn't appear here, I'm going to fix this.

File: src/ide/code_actions/fill_trait_members.rs [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+/// Generates a completion adding all trait members that have not yet been specified.
+/// Functions are added with empty bodies, consts with placeholder values.
+pub fn fill_trait_members(
+    db: &AnalysisDatabase,
+    node: SyntaxNode,
+    params: &CodeActionParams,
+) -> Option<CodeAction> {
+    let file = db.find_module_file_containing_node(&node)?.file_id(db).ok()?;
+
+    let item_impl_node = db.first_ancestor_of_kind(node, SyntaxKind::ItemImpl)?;
```
Comment (integraledelebesgue) [inline src/ide/code_actions/fill_trait_members.rs:31]: @piotmag769 Note: Only `impl`'s name is affected by the diagnostics and underlined in the `IDE`. Therefore, code action can be invoked only on the `impl` keyword, it's not available inside the body.

**Comment by integraledelebesgue** [comment]
Note: this action will not work until changes to the compiler from [this PR](https://reviewable.io/reviews/starkware-libs/cairo/6690) are not merged.

File: src/ide/code_actions/fill_trait_members.rs [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+            Some(param.id.name(db).map_or_else(
+                // concrete trait used only as a constraint
+                || format!("+{trait_name}{generic_arguments_bracket}"),
+                // concrete trait with explicit impl
+                |name| format!("impl {name}: {trait_name}{generic_arguments_bracket}"),
+            ))
+        }
+
+        other => Some(other.as_arg(db).format(db)),
```
Comment (piotmag769) [inline src/ide/code_actions/fill_trait_members.rs:237]: `NegImpl` is not allowed here no?

File: src/ide/code_actions/fill_trait_members.rs [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+    let nopanic = if !signature.panicable { Some(String::from("nopanic")) } else { None };
+
+    let body: Option<String> = Some(String::from("{}"));
+
+    Some([title, return_type, implicits, nopanic, body].into_iter().flatten().join(" "))
+}
+
+/// Formats [`GenericParam`] to be used as a free-standing type parameter
+/// (not belonging to the trait) in a generic function's declaration.
```
Comment (piotmag769) [inline src/ide/code_actions/fill_trait_members.rs:201]: That's not always true, sometimes you substitute trait generic types with concrete types, no?

Comment (integraledelebesgue) [reply src/ide/code_actions/fill_trait_members.rs:201]: Indeed, docstring is inaccurate. This function uses substitution rewriter on impl-related parameters but keeps "free" generics unchanged, so it's functionally correct

File: tests/test_data/code_actions/fill_trait_members.txt (hunk 1 above)
Comment (integraledelebesgue) [reply tests/test_data/code_actions/fill_trait_members.txt]: Done :)

File: src/ide/code_actions/fill_trait_members.rs (hunk 2 above)
Comment (piotmag769) [reply src/ide/code_actions/fill_trait_members.rs:31]: Thanks for the explanation!

File: src/ide/code_actions/fill_trait_members.rs [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+            let concrete_trait = param.concrete_trait.ok()?;
+            let trait_name = concrete_trait.name(db);
+            let trait_generic_arguments = concrete_trait.generic_args(db);
+
+            let generic_arguments_bracket = if trait_generic_arguments.is_empty() {
+                String::new()
+            } else {
+                let formatted_arguments = trait_generic_arguments
+                    .into_iter()
+                    .map(|argument| generic_argument_code(db, argument, type_concretizer))
```
Comment (piotmag769) [inline src/ide/code_actions/fill_trait_members.rs:227]: [nit] If all generics were substituted with concrete traits here (e.g. `+Drop<T>` with `+Drop<felt252>`) cannot we just drop the trait contraint? Or does cairo not allow it? It is really a nitpick, you can create an issue for that

File: src/ide/code_actions/fill_trait_members.rs [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+    let mut type_concretizer = SubstitutionRewriter { db, substitution: &substitution };
+
+    // Iterators borrowing SubstitutionRewriter mutably need intermediate collection.
+    let code = chain!(
+        trait_constants
+            .values()
+            .filter_map(|&id| constant_code(db, id, &mut type_concretizer))
+            .collect_vec()
+            .into_iter(),
+        trait_types.values().map(|id| format!("type {} = ();", id.name(db))),
```
Comment (Arcticae) [inline src/ide/code_actions/fill_trait_members.rs:79]: I would put types above the constants i think

File: src/ide/code_actions/fill_trait_members.rs [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        .join(", ");
+
+    let name = id.name(db);
+    let title = Some(format!("fn {name}{generic_parameters_bracket}({parameters})"));
+
+    let return_type = type_concretizer.rewrite(signature.return_type).ok()?;
+    let return_type =
+        if return_type.is_unit(db) { None } else { Some(format!("-> {}", return_type.format(db))) };
+
+    let implicits = match &signature.implicits[..] {
```
Comment (Arcticae) [inline src/ide/code_actions/fill_trait_members.rs:192]: do we need to generate those? 🤔

File: src/ide/code_actions/fill_trait_members.rs (hunk 5 above)
Comment (Arcticae) [reply src/ide/code_actions/fill_trait_members.rs:227]: It's safer to generate this more explicit version probably

File: src/ide/code_actions/mod.rs [hunk 8]
```diff
@@ -88,6 +89,9 @@ fn get_code_actions_for_diagnostics(
             "E0003" => fill_struct_fields::fill_struct_fields(db, node.clone(), params)
                 .map(|result| vec![result])
                 .unwrap_or_default(),
+            "E0004" => fill_trait_members::fill_trait_members(db, node.clone(), params)
```
Comment (Arcticae) [inline src/ide/code_actions/mod.rs:92]: What happens if it doesn't return anything? Nothing happens? 
I think that might be weird UX for the users:

1. You click "fill trait members"
2. Nothing happens 
3. You wait, or re-run and still nothing

If it's the case, i think it should be handled in a different PR (create an issue please 🙏 ) since it looks like other actions follow this pattern too

File: src/ide/code_actions/fill_trait_members.rs (hunk 5 above)
Comment (piotmag769) [reply src/ide/code_actions/fill_trait_members.rs:227]: Yep, it is not allowed in cairo, just checked it. Resolving

File: src/ide/code_actions/mod.rs (hunk 8 above)
Comment (piotmag769) [reply src/ide/code_actions/mod.rs:92]: >What happens if it doesn't return anything?

Then no code action is shown

```rs
                .map(|result| vec![result])
                .unwrap_or_default(),

**Review by Arcticae — APPROVED** [review]
